libraries/apollo-ast/src/commonMain/kotlin/com/apollographql/apollo3/ast/gqldirective.kt

```diff
@@ -74,3 +74,99 @@ fun List<GQLDirective>.optionalValue(schema: Schema?): Boolean? {
 
 @ApolloInternal
 fun List<GQLDirective>.findNonnull(schema: Schema) = any { schema.originalDirectiveName(it.name) == Schema.NONNULL }
+
+enum class CatchTo {
+  NO_CATCH,
+  THROW,
+  RESULT,
+  NULL
+}
+
+@ApolloInternal
+class Catch(val to: CatchTo, val level: Int?)
+
+private fun GQLDirectiveDefinition.getArgumentDefaultValue(argName: String): GQLValue? {
+  return arguments.firstOrNull { it.name == argName }?.defaultValue
+}
+
+private fun GQLDirective.getArgument(argName: String, schema: Schema): GQLValue? {
+  val directiveDefinition: GQLDirectiveDefinition = schema.directiveDefinitions.get(name)!!
+  val argument = arguments.firstOrNull { it.name == argName }
+  if (argument == null) {
+    return directiveDefinition.getArgumentDefaultValue(argName)
+  }
+  return argument.value
+}
+
+private fun GQLValue.toIntOrNull(): Int? {
+  return when (this) {
+    is GQLNullValue -> null
+    is GQLIntValue -> this.value.toIntOrNull()
+    else -> error("${sourceLocation}: expected Int value")
+  }
+}
+
+private fun GQLValue.toStringOrNull(): String? {
+  return when (this) {
+    is GQLNullValue -> null
+    is GQLStringValue -> this.value
+    else -> error("${sourceLocation}: expected String value")
+  }
+}
+
+private fun GQLValue?.toBoolean(): Boolean {
+  return when (this) {
+    is GQLBooleanValue -> this.value
+    else -> error("${this?.sourceLocation}: expected Boolean! value")
+  }
+}
+
+private fun GQLValue?.toCatchTo(): CatchTo {
+  return when (this) {
+    is GQLEnumValue -> when (this.value) {
+      "NULL" -> CatchTo.NULL
+      "RESULT" -> CatchTo.RESULT
+      "THROW" -> CatchTo.THROW
+      else -> error("Unknown CatchTo value: ${this.value}")
+    }
+
+    else -> error("${this?.sourceLocation}: expected CatchTo! value")
+  }
+}
+
+@ApolloInternal
+fun List<GQLDirective>.findCatchs(schema: Schema): List<Catch> {
+  return filter {
+    schema.originalDirectiveName(it.name) == Schema.CATCH
+  }.map {
+    val to = if (!it.getArgument("if", schema).toBoolean()) {
```

Comment:
Backward compat. Saves allocation of an adapter wrapper for every field.

Reply:
I meant, do we need an `if` argument on the directive? But I guess it's for consistency, as we have it in a lot of other directives.

Reply:
The alternative was 

```graphql
extend schema @catch(to: DO_NOT_CATCH)
```

Or maybe since this is supposed to be only used in schema extensions, use a separate directive?

```graphql
extend schema @throwOnFieldError
```

I'm not sure there's a lot of value in the other choices... (`NULL`, `RESULT`) although it's technically easy to do.

Or maybe 


```graphql
extend schema @enableCatch
```

?

Reply:
I didn't get the `schema` case before but now that I do, `if` looks the best to me.

Reply:
I changed this to not have any schema directive in https://github.com/apollographql/apollo-kotlin/pull/5405/commits/c141ad31ac7507254e63aec41c69aa1b283f934d. Instead, the error aware codegen is enabled if the `@catch` definition is present in the schema.